File: apps/desktop/src/contracts.ts
export const CONTRACT_VERSION = "1.0" as const;

export const MESSAGE_TYPES = ["request", "response", "progress", "error"] as const;
export const JOB_STATES = [
  "queued",
  "acquiring",
  "extracting",
  "ingesting",
  "validating",
  "staging",
  "publishing",
  "completed",
  "needs_review",
  "cancelled",
  "failed",
] as const;
export const SOURCE_FORMATS = ["pdf", "docx", "txt", "md"] as const;
export const ERROR_CATEGORIES = [
  "validation",
  "invalid_request",
  "unavailable",
  "timeout",
  "cancelled",
  "internal",
] as const;
export const REVIEW_SEVERITIES = ["info", "warning", "error"] as const;
export const PROVIDER_IDS = ["codex", "claude", "antigravity", "fake"] as const;
export const INGEST_STRATEGIES = ["forced_layout_ocr", "direct_text", "structured_docx"] as const;

export type MessageType = (typeof MESSAGE_TYPES)[number];
export type JobState = (typeof JOB_STATES)[number];
export type SourceFormat = (typeof SOURCE_FORMATS)[number];
export type ErrorCategory = (typeof ERROR_CATEGORIES)[number];
export type ReviewSeverity = (typeof REVIEW_SEVERITIES)[number];
export type ProviderId = (typeof PROVIDER_IDS)[number];
export type IngestStrategy = (typeof INGEST_STRATEGIES)[number];

export interface IpcEnvelope<TPayload extends Record<string, unknown> = Record<string, unknown>> {
  protocol_version: typeof CONTRACT_VERSION;
  message_type: MessageType;
  request_id: string;
  wiki_id?: string;
  job_id?: string;
  payload: TPayload;
}

export interface WikiRegistration {
  schema_version: typeof CONTRACT_VERSION;
  wiki_id: string;
  display_name: string;
  canonical_root: string;
  note_language: string;
  created_at: string;
  last_opened_at: string;
}

export interface WikiSettings {
  schema_version: typeof CONTRACT_VERSION;
  wiki_id: string;
  output_root: string;
  note_language: string;
  provider_id: ProviderId;
  ocr_language: string;
  open_in_obsidian_after_publish?: boolean;
}

export interface JobCheckpoint {
  state: JobState;
  completed_at: string;
  artifact_path?: string;
}

export interface JobRecord {
  schema_version: typeof CONTRACT_VERSION;
  job_id: string;
  wiki_id: string;
  state: JobState;
  stage_progress: number;
  created_at: string;
  updated_at: string;
  checkpoints: JobCheckpoint[];
}

export interface SourceManifestEntry {
  schema_version: typeof CONTRACT_VERSION;
  source_id: string;
  original_name: string;
  source_format: SourceFormat;
  content_sha256: string;
  byte_size: number;
  ingest_strategy: IngestStrategy;
}

export interface ExtractedPage {
  page_number: number;
  markdown: string;
  layout_confidence?: number;
}

export interface ExtractionArtifact {
  schema_version: typeof CONTRACT_VERSION;
  source_id: string;
  extractor: string;
  pages: ExtractedPage[];
  plain_text: string;
}

export interface WikiWrite {
  relative_path: string;
  content_sha256: string;
  markdown: string;
}

export interface WikiTransaction {
  schema_version: typeof CONTRACT_VERSION;
  transaction_id: string;
  wiki_id: string;
  writes: WikiWrite[];
  deletes: string[];
}

export interface ReviewItem {
  schema_version: typeof CONTRACT_VERSION;
  review_id: string;
  job_id: string;
  severity: ReviewSeverity;
  code: string;
  message: string;
  status: "open" | "accepted" | "rejected" | "resolved";
  relative_path?: string;
}

export interface PublicationResult {
  schema_version: typeof CONTRACT_VERSION;
  transaction_id: string;
  published_at: string;
  written_paths: string[];
  backup_created: boolean;
}

File: crates/app-core/src/contracts.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

pub const CONTRACT_VERSION: &str = "1.0";

pub const MESSAGE_TYPES: &[&str] = &["request", "response", "progress", "error"];
pub const JOB_STATES: &[&str] = &[
    "queued",
    "acquiring",
    "extracting",
    "ingesting",
    "validating",
    "staging",
    "publishing",
    "completed",
    "needs_review",
    "cancelled",
    "failed",
];
pub const SOURCE_FORMATS: &[&str] = &["pdf", "docx", "txt", "md"];
pub const ERROR_CATEGORIES: &[&str] = &[
    "validation",
    "invalid_request",
    "unavailable",
    "timeout",
    "cancelled",
    "internal",
];
pub const REVIEW_SEVERITIES: &[&str] = &["info", "warning", "error"];
pub const PROVIDER_IDS: &[&str] = &["codex", "claude", "antigravity", "fake"];
pub const INGEST_STRATEGIES: &[&str] = &["forced_layout_ocr", "direct_text", "structured_docx"];

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MessageType {
    Request,
    Response,
    Progress,
    Error,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum JobState {
    Queued,
    Acquiring,
    Extracting,
    Ingesting,
    Validating,
    Staging,
    Publishing,
    Completed,
    NeedsReview,
    Cancelled,
    Failed,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SourceFormat {
    Pdf,
    Docx,
    Txt,
    Md,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ErrorCategory {
    Validation,
    InvalidRequest,
    Unavailable,
    Timeout,
    Cancelled,
    Internal,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ReviewSeverity {
    Info,
    Warning,
    Error,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ProviderId {
    Codex,
    Claude,
    Antigravity,
    Fake,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum IngestStrategy {
    ForcedLayoutOcr,
    DirectText,
    StructuredDocx,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct IpcEnvelope {
    pub protocol_version: String,
    pub message_type: MessageType,
    pub request_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub wiki_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub job_id: Option<String>,
    pub payload: Value,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct WikiRegistration {
    pub schema_version: String,
    pub wiki_id: String,
    pub display_name: String,
    pub canonical_root: String,
    pub note_language: String,
    pub created_at: String,
    pub last_opened_at: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct WikiSettings {
    pub schema_version: String,
    pub wiki_id: String,
    pub output_root: String,
    pub note_language: String,
    pub provider_id: ProviderId,
    pub ocr_language: String,
    #[serde(default)]
    pub open_in_obsidian_after_publish: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct JobCheckpoint {
    pub state: JobState,
    pub completed_at: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub artifact_path: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct JobRecord {
    pub schema_version: String,
    pub job_id: String,
    pub wiki_id: String,
    pub state: JobState,
    pub stage_progress: f64,
    pub created_at: String,
    pub updated_at: String,
    pub checkpoints: Vec<JobCheckpoint>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct SourceManifestEntry {
    pub schema_version: String,
    pub source_id: String,
    pub original_name: String,
    pub source_format: SourceFormat,
    pub content_sha256: String,
    pub byte_size: u64,
    pub ingest_strategy: IngestStrategy,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ExtractedPage {
    pub page_number: u32,
    pub markdown: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub layout_confidence: Option<f64>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ExtractionArtifact {
    pub schema_version: String,
    pub source_id: String,
    pub extractor: String,
    pub pages: Vec<ExtractedPage>,
    pub plain_text: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct WikiWrite {
    pub relative_path: String,
    pub content_sha256: String,
    pub markdown: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct WikiTransaction {
    pub schema_version: String,
    pub transaction_id: String,
    pub wiki_id: String,
    pub writes: Vec<WikiWrite>,
    pub deletes: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ReviewStatus {
    Open,
    Accepted,
    Rejected,
    Resolved,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ReviewItem {
    pub schema_version: String,
    pub review_id: String,
    pub job_id: String,
    pub severity: ReviewSeverity,
    pub code: String,
    pub message: String,
    pub status: ReviewStatus,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub relative_path: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct PublicationResult {
    pub schema_version: String,
    pub transaction_id: String,
    pub published_at: String,
    pub written_paths: Vec<String>,
    pub backup_created: bool,
}
